# Problem statement

Can we infer important COVID-19 public health risk factors from outdated data? In many countries census and other survey data may be incomplete or out of date. This challenge is to develop a proof-of-concept for how machine learning can help governments more accurately map COVID-19 risk in 2020 using old data, without requiring a new costly, risky, and time-consuming on-the-ground survey.

The 2011 census gives us valuable information for determining who might be most vulnerable to COVID-19 in South Africa. However, the data is nearly 10 years old, and we expect that some key indicators will have changed in that time. Building an up-to-date map showing where the most vulnerable are located will be a key step in responding to the disease. A mapping effort like this requires bringing together many different inputs and tools. For this competition, we’re starting small. Can we infer important risk factors from more readily available data?

The task is to predict the percentage of households that fall into a particularly vulnerable bracket - large households who must leave their homes to fetch water - using 2011 South African census data. Solving this challenge will show that with machine learning it is possible to use easy-to-measure stats to identify areas most at risk even in years when census data is not collected.

The challenge can be found [here](https://zindi.africa/hackathons/south-african-covid-19-vulnerability-map/)

My team's winning solution can be found [here](https://github.com/KuriaDan/Position_2_solution_for_the__African-COVID-19__zindi_hackathon/blob/master/Position_2_(Team_SomeC)_solution_for_the___African_COVID_19__zindi_hackathon.ipynb)

In [115]:
# Importing data wrangling and machine Learning packages used.
import pandas as pd
import numpy as np
import seaborn as sns
sns.set_style('darkgrid')
import matplotlib.pyplot as plt

import requests
from io import StringIO 
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from lightgbm import LGBMRegressor
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV, RandomizedSearchCV, KFold
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Created links to shared files via google drive
#
train = 'https://drive.google.com/file/d/1_wLi9i-pUk6Kaizjb5i6-Pd0Vi7L1d-E/view?usp=sharing'
test = 'https://drive.google.com/file/d/1OeT53v7tZLnB71n1j4r6KFbtrouz38ej/view?usp=sharing'

# Created a function to read a csv file shared via google and return a dataframe
#
def read_csv(url):
    url = 'https://drive.google.com/uc?export=download&id=' + url.split('/')[-2]
    csv_raw = requests.get(url).text
    csv = StringIO(csv_raw)
    df = pd.read_csv(csv)
    return df

# Creating training and testing datataframes
#
train = read_csv(train)
test = read_csv(test)


print('Shape of the train data is {} while shape of Test data is {}'.format(train.shape, test.shape))


Shape of the train data is (3174, 50) while shape of Test data is (1102, 49)


In [3]:
# Previewing the first 5 observations of the train data.
train.head()

ward  total_households  total_individuals  target_pct_vunerable  \
0  9D9BEUB       13569.97801        39024.03083              1.739801   
1  RERH3XM       13593.88256        32879.94646              1.993381   
2  GJWA3BO        2698.30050         8261.71093              0.038340   
3  GW6VRB7        1965.96698         9069.77694              7.411006   
4  CN9JZJC        1852.07821         9908.84404              8.223907   

      dw_00     dw_01     dw_02     dw_03     dw_04     dw_05  ...    lan_11  \
0  0.354806  0.004781  0.170498  0.007861  0.006796  0.023082  ...  0.001723   
1  0.262140  0.002066  0.077669  0.000597  0.000977  0.003759  ...  0.041982   
2  0.728101  0.002292  0.051663  0.008209  0.001444  0.005089  ...  0.014786   
3  0.613206  0.138127  0.004999  0.002704  0.000000  0.002788  ...  0.003087   
4  0.797462  0.120830  0.060677  0.001155  0.001042  0.000508  ...  0.000117   

     lan_12  lan_13    lan_14     pg_00     pg_01     pg_02     pg_03  \
0  0.027796       0  0.026727  0.524666  0.041460  0.394550  0.029031   
1  0.021312       0  0.015754  0.993900  0.001829  0.000854  0.000586   
2  0.005827       0  0.049384  0.990007  0.003582  0.002546  0.003201   
3  0.003136       0  0.013562  0.928682  0.004464  0.009854  0.054701   
4  0.001471       0  0.000000  0.994072  0.000476  0.004033  0.000704   

      pg_04    lgt_00  
0  0.010292  0.599259  
1  0.002832  0.699136  
2  0.000663  0.972315  
3  0.002299  0.735333  
4  0.000714  0.646964  

[5 rows x 50 columns]

In [4]:
# Previewing the forst 5 observations of the test data.
test.head()

ward  total_households  total_individuals     dw_00     dw_01     dw_02  \
0  00A93O0        2514.04172         8037.06199  0.671962  0.061074  0.101516   
1  01V40KO        1829.71944         7494.86917  0.542857  0.390850  0.039623   
2  01ZHVW0        1721.99009         5847.43254  0.968071  0.003525  0.001431   
3  028YQJX       12388.02257        52211.66859  0.397274  0.004734  0.001068   
4  02G8W11        1754.25716         4266.57609  0.815197  0.016697  0.078571   

      dw_03     dw_04     dw_05     dw_06  ...    lan_11    lan_12  lan_13  \
0  0.006891  0.021287  0.015615  0.035974  ...  0.000420  0.011485       0   
1  0.000000  0.000000  0.000000  0.002633  ...  0.000000  0.000160       0   
2  0.000000  0.000000  0.000000  0.018546  ...  0.000631  0.006504       0   
3  0.000279  0.000799  0.197760  0.005227  ...  0.000593  0.008330       0   
4  0.008975  0.047882  0.015398  0.007264  ...  0.000234  0.016846       0   

     lan_14     pg_00     pg_01     pg_02     pg_03     pg_04    lgt_00  
0  0.063705  0.979272  0.009926  0.003224  0.002697  0.004882  0.521007  
1  0.000000  0.998938  0.000296  0.000301  0.000305  0.000160  0.693716  
2  0.000171  0.955895  0.012170  0.000000  0.030012  0.001923  0.933954  
3  0.000000  0.293629  0.694779  0.003238  0.000628  0.007726  0.921470  
4  0.178598  0.110861  0.054114  0.006062  0.823495  0.005470  0.996227  

[5 rows x 49 columns]

In [137]:
# While quickly glancing on the data, some columns appear to only have 0s. 
# Checking one such column to find the number of unique values in the column.
train.lan_13.nunique()

AttributeError: 'DataFrame' object has no attribute 'lan_13'

In [8]:
# Looping through the train data to drop those columns.

for col in train.columns:
    if train[col].nunique() == 1:
        train = train.drop(col, axis=1)

        

In [7]:
train.head()

ward  total_households  total_individuals  target_pct_vunerable  \
0  9D9BEUB       13569.97801        39024.03083              1.739801   
1  RERH3XM       13593.88256        32879.94646              1.993381   
2  GJWA3BO        2698.30050         8261.71093              0.038340   
3  GW6VRB7        1965.96698         9069.77694              7.411006   
4  CN9JZJC        1852.07821         9908.84404              8.223907   

      dw_00     dw_01     dw_02     dw_03     dw_04     dw_05  ...    lan_10  \
0  0.354806  0.004781  0.170498  0.007861  0.006796  0.023082  ...  0.000787   
1  0.262140  0.002066  0.077669  0.000597  0.000977  0.003759  ...  0.005768   
2  0.728101  0.002292  0.051663  0.008209  0.001444  0.005089  ...  0.002755   
3  0.613206  0.138127  0.004999  0.002704  0.000000  0.002788  ...  0.001135   
4  0.797462  0.120830  0.060677  0.001155  0.001042  0.000508  ...  0.000236   

     lan_11    lan_12    lan_14     pg_00     pg_01     pg_02     pg_03  \
0  0.001723  0.027796  0.026727  0.524666  0.041460  0.394550  0.029031   
1  0.041982  0.021312  0.015754  0.993900  0.001829  0.000854  0.000586   
2  0.014786  0.005827  0.049384  0.990007  0.003582  0.002546  0.003201   
3  0.003087  0.003136  0.013562  0.928682  0.004464  0.009854  0.054701   
4  0.000117  0.001471  0.000000  0.994072  0.000476  0.004033  0.000704   

      pg_04    lgt_00  
0  0.010292  0.599259  
1  0.002832  0.699136  
2  0.000663  0.972315  
3  0.002299  0.735333  
4  0.000714  0.646964  

[5 rows x 47 columns]

In [136]:
# Checking for null values
if train.isnull().sum().sum() == 0:
    print('no NaN found')

no NaN found


## EDA

In [11]:
train.target_pct_vunerable.describe()

count    3174.000000
mean       10.258947
std        12.847192
min         0.000000
25%         0.429761
50%         3.715314
75%        17.762604
max       100.000000
Name: target_pct_vunerable, dtype: float64

  ### Baseline CatBoost model.


In [139]:
# Training a baseline LGBM model
X=train.drop(["ward",'target_pct_vunerable'],axis =1)
y = train.target_pct_vunerable

x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.20)
est = LGBMRegressor().fit(x_train, y_train)
y_pred = est.predict(x_test)
np.sqrt(mean_squared_error(y_pred, y_test))


5.819721389412163

In [143]:
# Training a baseline LGBM model
X=train.drop(["ward",'target_pct_vunerable'],axis =1)
y = train.target_pct_vunerable

x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=90)
est = LGBMRegressor().fit(x_train, y_train)
y_pred = est.predict(x_test)
np.sqrt(mean_squared_error(y_pred, y_test))


5.600945528134281

In [18]:
print(LGBMRegressor())

LGBMRegressor(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
              importance_type='split', learning_rate=0.1, max_depth=-1,
              min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
              n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
              random_state=None, reg_alpha=0.0, reg_lambda=0.0, silent=True,
              subsample=1.0, subsample_for_bin=200000, subsample_freq=0)


### Hyperparameters

> boosting_type - LGBM can run different Gradient Boosting methods, from traditional GBDT to DART(dropout) and GOSS(gradient based one side sampling)

> num_leaves - Sets the maximum number of leaves each weak learner has.

> Subsample - also called bagging fraction. specifies the percentage of rows used per tree building iteration.

> max_depth - maximum depth of a tree.

> num_iterations - number of boosting iterations. 

> Learning rate - determines the step size at each interation while moving towards a minimum of a loss function.

## Approach 1: 

### Manually tuning by trial and error.

In [55]:
# Tuning the boosting 
est = LGBMRegressor(boosting_type='dart').fit(x_train, y_train)
y_pred = est.predict(x_test)
np.sqrt(mean_squared_error(y_pred, y_test))


6.13047631213017

In [56]:
est = LGBMRegressor(boosting_type='goss').fit(x_train, y_train)
y_pred = est.predict(x_test)
np.sqrt(mean_squared_error(y_pred, y_test))


266482.3573341219

## Approach 2: 

### Manually tuning using cross validation

In [81]:
%%time
# Here, we increase the number of trees while reducing the learning rate with the same factor.
# First, we try with a five fold cross validation.

scores = cross_val_score(LGBMRegressor(n_estimators=1000, learning_rate=0.01), X, y, cv=5, scoring='neg_root_mean_squared_error')
print(scores)
print(np.mean(scores))

[-6.19706013 -6.74925307 -5.90050262 -5.89502568 -5.48076016]
-6.044520331167662
Wall time: 44.3 s


In [82]:
%%time
# we keep everything constant but change the number of folds from 5 to 7.
scores = cross_val_score(LGBMRegressor(n_estimators=1000, learning_rate=0.01), X, y, cv=7, scoring='neg_root_mean_squared_error')
print(scores)
print(np.mean(scores))

[-6.0357947  -6.21211195 -6.85421203 -6.01496085 -5.75760133 -5.42536874
 -5.64013889]
-5.991455498711916
Wall time: 45.5 s


### Automating above process with a for loop

In [144]:
np.arange(0.01, 0.1, 0.02)

array([0.01, 0.03, 0.05, 0.07, 0.09])

In [85]:
%%time
trees = [900, 1000]
etas = np.arange(0.01, 0.1, 0.02)
leaves=[25, 30, 35]
average_scores = {}

for tree in trees:
    for eta in etas:
        for leaf in leaves:
            cv_scores=cross_val_score(LGBMRegressor(n_estimators=tree, learning_rate=eta, num_leaves=leaf), X, y, cv=7, scoring='neg_root_mean_squared_error')
            average_scores[str(tree)+'_'+str(eta)+'_'+str(leaf)] = np.average(cv_scores)

                
average_scores

Wall time: 21min 11s


{'900_0.01_25': -6.024943998262722,
 '900_0.01_30': -6.0033730629250615,
 '900_0.01_35': -6.000240447936209,
 '900_0.03_25': -5.99873906147097,
 '900_0.03_30': -5.998782841644078,
 '900_0.03_35': -5.991237253591733,
 '900_0.049999999999999996_25': -5.978094133503604,
 '900_0.049999999999999996_30': -6.027061729006964,
 '900_0.049999999999999996_35': -6.042119483239447,
 '900_0.06999999999999999_25': -6.079160782611086,
 '900_0.06999999999999999_30': -6.0199674358488755,
 '900_0.06999999999999999_35': -6.042249308664408,
 '900_0.08999999999999998_25': -6.041643675170066,
 '900_0.08999999999999998_30': -6.060179108707479,
 '900_0.08999999999999998_35': -6.0499729080135305,
 '1000_0.01_25': -6.023630889176753,
 '1000_0.01_30': -5.999268020470362,
 '1000_0.01_35': -5.995237830739831,
 '1000_0.03_25': -5.997244148162144,
 '1000_0.03_30': -5.999038051101979,
 '1000_0.03_35': -5.990207892424167,
 '1000_0.049999999999999996_25': -5.977333086455718,
 '1000_0.049999999999999996_30': -6.027365960

## Approach 3: GridSearchCV

In [145]:
dir(est)

['_Booster',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_base_doc',
 '_best_iteration',
 '_best_score',
 '_classes',
 '_estimator_type',
 '_evals_result',
 '_fobj',
 '_get_param_names',
 '_get_tags',
 '_more_tags',
 '_n_classes',
 '_n_features',
 '_objective',
 '_other_params',
 'best_iteration_',
 'best_score_',
 'booster_',
 'boosting_type',
 'class_weight',
 'colsample_bytree',
 'evals_result_',
 'feature_importances_',
 'fit',
 'get_params',
 'importance_type',
 'learning_rate',
 'max_depth',
 'min_child_samples',
 'min_child_weight',
 'min_split_gain',
 'n_estimators',
 'n_features_',
 'n_jobs',
 'num_leaves',
 'objective',


In [102]:
%%time
# Gridsearch does exactly the same thing as the for loop above but in a single line of code.

est = GridSearchCV(LGBMRegressor(), {
    'n_estimators': [900, 1000],
    'learning_rate': np.arange(0.01, 0.1, 0.02),
    'num_leaves': [25,30,35]
}, cv=7, scoring='neg_root_mean_squared_error')
est.fit(X, y)
est.cv_results_

Wall time: 20min 14s


{'mean_fit_time': array([4.39750886, 5.28376647, 6.2964249 , 5.3736215 , 5.97589145,
        9.8461131 , 5.03412444, 5.09619073, 6.07820521, 5.50162411,
        6.17924782, 7.33211609, 4.54206879, 4.99899602, 6.09905366,
        4.67715536, 5.47795354, 6.31255528, 4.1936357 , 6.02828632,
        8.22946838, 4.83653658, 5.6129539 , 6.36149624, 4.21140259,
        4.96741959, 6.15578423, 4.74268835, 5.51626086, 6.43970701]),
 'std_fit_time': array([0.24930933, 0.47002929, 0.51529774, 0.57432461, 0.27951155,
        0.74424484, 1.12396682, 0.28242381, 0.35414954, 0.5062751 ,
        0.43130033, 0.45773767, 0.21134533, 0.22771023, 0.3770536 ,
        0.09468852, 0.1360997 , 0.17045921, 0.09287023, 1.18877976,
        1.35297405, 0.43341485, 0.34997158, 0.13791134, 0.09008101,
        0.15058712, 0.51151248, 0.05894681, 0.11312743, 0.1669518 ]),
 'mean_score_time': array([0.01957556, 0.0297151 , 0.02857229, 0.02971567, 0.0274178 ,
        0.0306687 , 0.02672124, 0.02857137, 0.03428667, 0.03

In [103]:
results=pd.DataFrame(est.cv_results_)
results

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        4.397509      0.249309         0.019576        0.003847   
1        5.283766      0.470029         0.029715        0.011877   
2        6.296425      0.515298         0.028572        0.008398   
3        5.373621      0.574325         0.029716        0.009285   
4        5.975891      0.279512         0.027418        0.003944   
5        9.846113      0.744245         0.030669        0.002841   
6        5.034124      1.123967         0.026721        0.003494   
7        5.096191      0.282424         0.028571        0.003957   
8        6.078205      0.354150         0.034287        0.008241   
9        5.501624      0.506275         0.031801        0.004397   
10       6.179248      0.431300         0.027430        0.003959   
11       7.332116      0.457738         0.042270        0.010083   
12       4.542069      0.211345         0.027430        0.008397   
13       4.998996      0.227710         0.028573        0.003959   
14       6.099054      0.377054         0.030859        0.007917   
15       4.677155      0.094689         0.026287        0.003614   
16       5.477954      0.136100         0.030858        0.005110   
17       6.312555      0.170459         0.034435        0.005550   
18       4.193636      0.092870         0.026287        0.009285   
19       6.028286      1.188780         0.029600        0.003552   
20       8.229468      1.352974         0.035641        0.004235   
21       4.836537      0.433415         0.025144        0.005111   
22       5.612954      0.349972         0.029717        0.005599   
23       6.361496      0.137911         0.035429        0.003958   
24       4.211403      0.090081         0.024001        0.000001   
25       4.967420      0.150587         0.025144        0.005111   
26       6.155784      0.511512         0.031945        0.004279   
27       4.742688      0.058947         0.027430        0.003959   
28       5.516261      0.113127         0.030860        0.002799   
29       6.439707      0.166952         0.036569        0.007226   

   param_learning_rate param_n_estimators param_num_leaves  \
0                 0.01                900               25   
1                 0.01                900               30   
2                 0.01                900               35   
3                 0.01               1000               25   
4                 0.01               1000               30   
5                 0.01               1000               35   
6                 0.03                900               25   
7                 0.03                900               30   
8                 0.03                900               35   
9                 0.03               1000               25   
10                0.03               1000               30   
11                0.03               1000               35   
12                0.05                900               25   
13                0.05                900               30   
14                0.05                900               35   
15                0.05               1000               25   
16                0.05               1000               30   
17                0.05               1000               35   
18                0.07                900               25   
19                0.07                900               30   
20                0.07                900               35   
21                0.07               1000               25   
22                0.07               1000               30   
23                0.07               1000               35   
24                0.09                900               25   
25                0.09                900               30   
26                0.09                900               35   
27                0.09               1000               25   
28                0.09               1000               30   
29                0.09             

In [104]:
results[['param_n_estimators', 'param_learning_rate', 'param_num_leaves','mean_test_score']]

param_n_estimators param_learning_rate param_num_leaves  mean_test_score
0                 900                0.01               25        -6.024944
1                 900                0.01               30        -6.003373
2                 900                0.01               35        -6.000240
3                1000                0.01               25        -6.023631
4                1000                0.01               30        -5.999268
5                1000                0.01               35        -5.995238
6                 900                0.03               25        -5.998739
7                 900                0.03               30        -5.998783
8                 900                0.03               35        -5.991237
9                1000                0.03               25        -5.997244
10               1000                0.03               30        -5.999038
11               1000                0.03               35        -5.990208
12                900                0.05               25        -5.978094
13                900                0.05               30        -6.027062
14                900                0.05               35        -6.042119
15               1000                0.05               25        -5.977333
16               1000                0.05               30        -6.027366
17               1000                0.05               35        -6.042738
18                900                0.07               25        -6.079161
19                900                0.07               30        -6.019967
20                900                0.07               35        -6.042249
21               1000                0.07               25        -6.080088
22               1000                0.07               30        -6.020696
23               1000                0.07               35        -6.042756
24                900                0.09               25        -6.041644
25                900                0.09               30        -6.060179
26                900                0.09               35        -6.049973
27               1000                0.09               25        -6.042486
28               1000                0.09               30        -6.060407
29               1000                0.09               35        -6.050062

In [105]:
est.best_params_

{'learning_rate': 0.049999999999999996, 'n_estimators': 1000, 'num_leaves': 25}

In [106]:
est.best_score_

-5.977333086455718

## Approach 4

### Random Search

In [132]:
%%time
# Random search reduces the number of iterations by using  random combination of the hyperparameter grid.
# It's therefore inexpensive computationally.
est = RandomizedSearchCV(LGBMRegressor(), {
    'n_estimators': [900, 1000],
    'learning_rate': np.arange(0.01, 0.1, 0.02),
    'num_leaves': [25,30,35]
}, cv=7, scoring='neg_root_mean_squared_error', n_iter=8)
est.fit(X, y)
est.cv_results_

Wall time: 5min 50s


{'mean_fit_time': array([8.31200695, 6.82716094, 6.68910142, 5.91374772, 5.7614413 ,
        4.35420694, 6.46812187, 4.51092979]),
 'std_fit_time': array([1.43794783, 0.49508203, 0.46940929, 0.46553177, 0.14452858,
        0.09432378, 0.10913395, 0.06095054]),
 'mean_score_time': array([0.04060701, 0.04046375, 0.03889765, 0.02878199, 0.0286384 ,
        0.02379496, 0.0287822 , 0.02208533]),
 'std_score_time': array([0.01611359, 0.0131857 , 0.0107679 , 0.00187963, 0.00254074,
        0.00098704, 0.0030359 , 0.00063705]),
 'param_num_leaves': masked_array(data=[30, 30, 35, 30, 35, 25, 35, 25],
              mask=[False, False, False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'param_n_estimators': masked_array(data=[1000, 1000, 1000, 1000, 900, 900, 1000, 900],
              mask=[False, False, False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'param_learning_rate': masked_array(data=[0.01, 0.03, 

In [110]:
pd.DataFrame(est.cv_results_)[['param_n_estimators', 'param_learning_rate', 'param_num_leaves','mean_test_score']]

param_n_estimators param_learning_rate param_num_leaves  mean_test_score
0               1000                0.03               35        -5.990208
1                900                0.07               25        -6.079161
2               1000                0.07               25        -6.080088
3               1000                0.09               35        -6.050062
4               1000                0.01               35        -5.995238
5                900                0.03               30        -5.998783
6                900                0.09               25        -6.041644
7                900                0.05               25        -5.978094
8               1000                0.05               25        -5.977333
9                900                0.05               30        -6.027062

In [112]:
est.best_params_

{'num_leaves': 35, 'n_estimators': 1000, 'learning_rate': 0.03}

In [113]:
est.best_score_

-5.990207892424167

## Finally...

In [120]:
# Creating the model with the best parameters using kfold cross-validation.
X=train.drop(["ward",'target_pct_vunerable'],axis =1)
y = train.target_pct_vunerable


score=[]


fold=KFold(n_splits=7)
for train_index, test_index in fold.split(X,y):
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]
    m2  = LGBMRegressor(n_estimators=1000, learning_rate=0.05, number_of_leaves=25, silent=False)
    m2.fit(X_train,y_train,eval_set=[(X_train,y_train),(X_test, y_test)])
    preds=m2.predict(X_test)
    print("err: ",np.sqrt(mean_squared_error(y_test,preds)))
    score.append(np.sqrt(mean_squared_error(y_test,preds)))


[1]	training's l2: 151.2	valid_1's l2: 161.329
[2]	training's l2: 140.155	valid_1's l2: 150.21
[3]	training's l2: 130.014	valid_1's l2: 140.005
[4]	training's l2: 120.752	valid_1's l2: 131.169
[5]	training's l2: 112.365	valid_1's l2: 123.188
[6]	training's l2: 104.736	valid_1's l2: 115.661
[7]	training's l2: 97.8194	valid_1's l2: 108.84
[8]	training's l2: 91.5163	valid_1's l2: 102.769
[9]	training's l2: 85.5958	valid_1's l2: 97.0565
[10]	training's l2: 80.2592	valid_1's l2: 91.9018
[11]	training's l2: 75.4213	valid_1's l2: 87.5075
[12]	training's l2: 70.8903	valid_1's l2: 83.3745
[13]	training's l2: 66.7821	valid_1's l2: 79.365
[14]	training's l2: 63.0291	valid_1's l2: 75.9629
[15]	training's l2: 59.5148	valid_1's l2: 72.4686
[16]	training's l2: 56.3858	valid_1's l2: 69.6038
[17]	training's l2: 53.5181	valid_1's l2: 66.9955
[18]	training's l2: 50.8492	valid_1's l2: 64.7802
[19]	training's l2: 48.2901	valid_1's l2: 62.3337
[20]	training's l2: 46.0285	valid_1's l2: 60.3501
[21]	training'

[172]	training's l2: 4.45965	valid_1's l2: 36.4824
[173]	training's l2: 4.42632	valid_1's l2: 36.4777
[174]	training's l2: 4.39052	valid_1's l2: 36.4593
[175]	training's l2: 4.35508	valid_1's l2: 36.4383
[176]	training's l2: 4.31704	valid_1's l2: 36.434
[177]	training's l2: 4.28184	valid_1's l2: 36.4222
[178]	training's l2: 4.24684	valid_1's l2: 36.3546
[179]	training's l2: 4.20963	valid_1's l2: 36.3566
[180]	training's l2: 4.18024	valid_1's l2: 36.3656
[181]	training's l2: 4.14969	valid_1's l2: 36.354
[182]	training's l2: 4.11472	valid_1's l2: 36.3658
[183]	training's l2: 4.06839	valid_1's l2: 36.3615
[184]	training's l2: 4.04139	valid_1's l2: 36.3646
[185]	training's l2: 4.00339	valid_1's l2: 36.3725
[186]	training's l2: 3.96999	valid_1's l2: 36.3981
[187]	training's l2: 3.94347	valid_1's l2: 36.3886
[188]	training's l2: 3.91318	valid_1's l2: 36.3855
[189]	training's l2: 3.87756	valid_1's l2: 36.4083
[190]	training's l2: 3.85023	valid_1's l2: 36.4063
[191]	training's l2: 3.81906	vali

[345]	training's l2: 1.37098	valid_1's l2: 36.3256
[346]	training's l2: 1.36265	valid_1's l2: 36.3202
[347]	training's l2: 1.35616	valid_1's l2: 36.3068
[348]	training's l2: 1.34965	valid_1's l2: 36.2965
[349]	training's l2: 1.34045	valid_1's l2: 36.2936
[350]	training's l2: 1.33303	valid_1's l2: 36.2944
[351]	training's l2: 1.32362	valid_1's l2: 36.2884
[352]	training's l2: 1.31513	valid_1's l2: 36.2833
[353]	training's l2: 1.3082	valid_1's l2: 36.2859
[354]	training's l2: 1.30022	valid_1's l2: 36.272
[355]	training's l2: 1.29354	valid_1's l2: 36.2657
[356]	training's l2: 1.28553	valid_1's l2: 36.2661
[357]	training's l2: 1.27862	valid_1's l2: 36.2676
[358]	training's l2: 1.27022	valid_1's l2: 36.2767
[359]	training's l2: 1.26078	valid_1's l2: 36.2786
[360]	training's l2: 1.25525	valid_1's l2: 36.2731
[361]	training's l2: 1.24848	valid_1's l2: 36.2812
[362]	training's l2: 1.24293	valid_1's l2: 36.2843
[363]	training's l2: 1.2366	valid_1's l2: 36.3064
[364]	training's l2: 1.2276	valid_

[512]	training's l2: 0.562043	valid_1's l2: 36.3107
[513]	training's l2: 0.559326	valid_1's l2: 36.3182
[514]	training's l2: 0.556427	valid_1's l2: 36.3177
[515]	training's l2: 0.553608	valid_1's l2: 36.3163
[516]	training's l2: 0.55134	valid_1's l2: 36.3144
[517]	training's l2: 0.548502	valid_1's l2: 36.3171
[518]	training's l2: 0.545863	valid_1's l2: 36.3175
[519]	training's l2: 0.5428	valid_1's l2: 36.3155
[520]	training's l2: 0.540541	valid_1's l2: 36.3145
[521]	training's l2: 0.537392	valid_1's l2: 36.3179
[522]	training's l2: 0.534917	valid_1's l2: 36.3206
[523]	training's l2: 0.532145	valid_1's l2: 36.3186
[524]	training's l2: 0.529036	valid_1's l2: 36.3214
[525]	training's l2: 0.52648	valid_1's l2: 36.3236
[526]	training's l2: 0.523763	valid_1's l2: 36.3182
[527]	training's l2: 0.521221	valid_1's l2: 36.3132
[528]	training's l2: 0.519107	valid_1's l2: 36.3123
[529]	training's l2: 0.516367	valid_1's l2: 36.3099
[530]	training's l2: 0.514019	valid_1's l2: 36.3089
[531]	training's

[687]	training's l2: 0.25638	valid_1's l2: 36.2591
[688]	training's l2: 0.255208	valid_1's l2: 36.2591
[689]	training's l2: 0.253929	valid_1's l2: 36.2627
[690]	training's l2: 0.252937	valid_1's l2: 36.2624
[691]	training's l2: 0.252065	valid_1's l2: 36.2585
[692]	training's l2: 0.250989	valid_1's l2: 36.2624
[693]	training's l2: 0.250046	valid_1's l2: 36.2637
[694]	training's l2: 0.248844	valid_1's l2: 36.2622
[695]	training's l2: 0.248035	valid_1's l2: 36.2591
[696]	training's l2: 0.246841	valid_1's l2: 36.2567
[697]	training's l2: 0.245724	valid_1's l2: 36.2542
[698]	training's l2: 0.244812	valid_1's l2: 36.2552
[699]	training's l2: 0.243657	valid_1's l2: 36.2552
[700]	training's l2: 0.242592	valid_1's l2: 36.2537
[701]	training's l2: 0.2415	valid_1's l2: 36.2532
[702]	training's l2: 0.240271	valid_1's l2: 36.2524
[703]	training's l2: 0.239142	valid_1's l2: 36.2548
[704]	training's l2: 0.238149	valid_1's l2: 36.2557
[705]	training's l2: 0.237278	valid_1's l2: 36.2572
[706]	training'

[858]	training's l2: 0.129431	valid_1's l2: 36.246
[859]	training's l2: 0.128961	valid_1's l2: 36.2431
[860]	training's l2: 0.128444	valid_1's l2: 36.2444
[861]	training's l2: 0.127958	valid_1's l2: 36.2417
[862]	training's l2: 0.127533	valid_1's l2: 36.2451
[863]	training's l2: 0.126991	valid_1's l2: 36.2468
[864]	training's l2: 0.126566	valid_1's l2: 36.2443
[865]	training's l2: 0.126182	valid_1's l2: 36.2481
[866]	training's l2: 0.12564	valid_1's l2: 36.2504
[867]	training's l2: 0.125133	valid_1's l2: 36.2508
[868]	training's l2: 0.124654	valid_1's l2: 36.2502
[869]	training's l2: 0.124129	valid_1's l2: 36.2496
[870]	training's l2: 0.123668	valid_1's l2: 36.2528
[871]	training's l2: 0.12308	valid_1's l2: 36.2516
[872]	training's l2: 0.122596	valid_1's l2: 36.2491
[873]	training's l2: 0.122114	valid_1's l2: 36.2496
[874]	training's l2: 0.121657	valid_1's l2: 36.2508
[875]	training's l2: 0.121225	valid_1's l2: 36.2497
[876]	training's l2: 0.120641	valid_1's l2: 36.2502
[877]	training'

[20]	training's l2: 46.2316	valid_1's l2: 54.341
[21]	training's l2: 44.1083	valid_1's l2: 52.7989
[22]	training's l2: 42.1305	valid_1's l2: 51.4135
[23]	training's l2: 40.2641	valid_1's l2: 50.2628
[24]	training's l2: 38.5735	valid_1's l2: 49.3855
[25]	training's l2: 36.9737	valid_1's l2: 48.3293
[26]	training's l2: 35.4623	valid_1's l2: 47.2937
[27]	training's l2: 34.0929	valid_1's l2: 46.3424
[28]	training's l2: 32.8283	valid_1's l2: 45.7531
[29]	training's l2: 31.6487	valid_1's l2: 45.0632
[30]	training's l2: 30.5219	valid_1's l2: 44.4463
[31]	training's l2: 29.507	valid_1's l2: 43.8781
[32]	training's l2: 28.5709	valid_1's l2: 43.2644
[33]	training's l2: 27.6033	valid_1's l2: 42.7793
[34]	training's l2: 26.7486	valid_1's l2: 42.4521
[35]	training's l2: 25.9178	valid_1's l2: 42.1375
[36]	training's l2: 25.169	valid_1's l2: 41.74
[37]	training's l2: 24.4878	valid_1's l2: 41.4632
[38]	training's l2: 23.8011	valid_1's l2: 40.9733
[39]	training's l2: 23.1578	valid_1's l2: 40.6758
[40]	

[191]	training's l2: 3.80182	valid_1's l2: 36.4123
[192]	training's l2: 3.77071	valid_1's l2: 36.4165
[193]	training's l2: 3.74287	valid_1's l2: 36.3915
[194]	training's l2: 3.71745	valid_1's l2: 36.3803
[195]	training's l2: 3.69303	valid_1's l2: 36.3704
[196]	training's l2: 3.66336	valid_1's l2: 36.3618
[197]	training's l2: 3.63435	valid_1's l2: 36.3504
[198]	training's l2: 3.60553	valid_1's l2: 36.3476
[199]	training's l2: 3.57744	valid_1's l2: 36.3251
[200]	training's l2: 3.54505	valid_1's l2: 36.3555
[201]	training's l2: 3.52086	valid_1's l2: 36.3507
[202]	training's l2: 3.49943	valid_1's l2: 36.3595
[203]	training's l2: 3.47216	valid_1's l2: 36.3463
[204]	training's l2: 3.44607	valid_1's l2: 36.3413
[205]	training's l2: 3.4213	valid_1's l2: 36.3315
[206]	training's l2: 3.3974	valid_1's l2: 36.3
[207]	training's l2: 3.36943	valid_1's l2: 36.3163
[208]	training's l2: 3.34544	valid_1's l2: 36.336
[209]	training's l2: 3.31658	valid_1's l2: 36.3053
[210]	training's l2: 3.29012	valid_1'

[356]	training's l2: 1.27714	valid_1's l2: 36.2605
[357]	training's l2: 1.27025	valid_1's l2: 36.259
[358]	training's l2: 1.26288	valid_1's l2: 36.2626
[359]	training's l2: 1.25718	valid_1's l2: 36.2472
[360]	training's l2: 1.25056	valid_1's l2: 36.2394
[361]	training's l2: 1.24406	valid_1's l2: 36.2427
[362]	training's l2: 1.23769	valid_1's l2: 36.2249
[363]	training's l2: 1.23276	valid_1's l2: 36.2239
[364]	training's l2: 1.22576	valid_1's l2: 36.213
[365]	training's l2: 1.21929	valid_1's l2: 36.2296
[366]	training's l2: 1.21219	valid_1's l2: 36.2249
[367]	training's l2: 1.20615	valid_1's l2: 36.2179
[368]	training's l2: 1.19959	valid_1's l2: 36.2339
[369]	training's l2: 1.19196	valid_1's l2: 36.2371
[370]	training's l2: 1.18427	valid_1's l2: 36.2277
[371]	training's l2: 1.1756	valid_1's l2: 36.24
[372]	training's l2: 1.16948	valid_1's l2: 36.2358
[373]	training's l2: 1.16308	valid_1's l2: 36.2279
[374]	training's l2: 1.15632	valid_1's l2: 36.2368
[375]	training's l2: 1.15048	valid_1

[535]	training's l2: 0.520333	valid_1's l2: 36.1331
[536]	training's l2: 0.517715	valid_1's l2: 36.1286
[537]	training's l2: 0.515726	valid_1's l2: 36.1275
[538]	training's l2: 0.513136	valid_1's l2: 36.1215
[539]	training's l2: 0.510817	valid_1's l2: 36.132
[540]	training's l2: 0.508324	valid_1's l2: 36.1283
[541]	training's l2: 0.505979	valid_1's l2: 36.1244
[542]	training's l2: 0.503758	valid_1's l2: 36.117
[543]	training's l2: 0.501908	valid_1's l2: 36.1138
[544]	training's l2: 0.49917	valid_1's l2: 36.1216
[545]	training's l2: 0.497207	valid_1's l2: 36.1166
[546]	training's l2: 0.495147	valid_1's l2: 36.1107
[547]	training's l2: 0.492716	valid_1's l2: 36.1296
[548]	training's l2: 0.490753	valid_1's l2: 36.1291
[549]	training's l2: 0.488362	valid_1's l2: 36.1266
[550]	training's l2: 0.486641	valid_1's l2: 36.1202
[551]	training's l2: 0.484268	valid_1's l2: 36.125
[552]	training's l2: 0.482084	valid_1's l2: 36.1168
[553]	training's l2: 0.479773	valid_1's l2: 36.1084
[554]	training's

[712]	training's l2: 0.248152	valid_1's l2: 36.1092
[713]	training's l2: 0.246987	valid_1's l2: 36.1045
[714]	training's l2: 0.245824	valid_1's l2: 36.1042
[715]	training's l2: 0.244844	valid_1's l2: 36.1028
[716]	training's l2: 0.243818	valid_1's l2: 36.1109
[717]	training's l2: 0.242786	valid_1's l2: 36.109
[718]	training's l2: 0.24193	valid_1's l2: 36.1047
[719]	training's l2: 0.241067	valid_1's l2: 36.1134
[720]	training's l2: 0.240094	valid_1's l2: 36.1144
[721]	training's l2: 0.239134	valid_1's l2: 36.1088
[722]	training's l2: 0.238099	valid_1's l2: 36.1081
[723]	training's l2: 0.237156	valid_1's l2: 36.1149
[724]	training's l2: 0.235873	valid_1's l2: 36.112
[725]	training's l2: 0.235037	valid_1's l2: 36.1116
[726]	training's l2: 0.234156	valid_1's l2: 36.1111
[727]	training's l2: 0.233217	valid_1's l2: 36.1206
[728]	training's l2: 0.232337	valid_1's l2: 36.1163
[729]	training's l2: 0.231563	valid_1's l2: 36.1149
[730]	training's l2: 0.230689	valid_1's l2: 36.1164
[731]	training'

[880]	training's l2: 0.132556	valid_1's l2: 36.1647
[881]	training's l2: 0.131943	valid_1's l2: 36.1661
[882]	training's l2: 0.131487	valid_1's l2: 36.1632
[883]	training's l2: 0.131046	valid_1's l2: 36.1612
[884]	training's l2: 0.130378	valid_1's l2: 36.164
[885]	training's l2: 0.129977	valid_1's l2: 36.1695
[886]	training's l2: 0.129423	valid_1's l2: 36.1703
[887]	training's l2: 0.128945	valid_1's l2: 36.1654
[888]	training's l2: 0.128448	valid_1's l2: 36.1671
[889]	training's l2: 0.128096	valid_1's l2: 36.1663
[890]	training's l2: 0.127708	valid_1's l2: 36.1741
[891]	training's l2: 0.127299	valid_1's l2: 36.171
[892]	training's l2: 0.126924	valid_1's l2: 36.1703
[893]	training's l2: 0.126475	valid_1's l2: 36.1722
[894]	training's l2: 0.126083	valid_1's l2: 36.1683
[895]	training's l2: 0.125662	valid_1's l2: 36.1689
[896]	training's l2: 0.125162	valid_1's l2: 36.1701
[897]	training's l2: 0.124722	valid_1's l2: 36.1669
[898]	training's l2: 0.124335	valid_1's l2: 36.1687
[899]	training

[42]	training's l2: 20.9587	valid_1's l2: 49.1337
[43]	training's l2: 20.4329	valid_1's l2: 48.9943
[44]	training's l2: 19.952	valid_1's l2: 48.9981
[45]	training's l2: 19.4978	valid_1's l2: 48.8439
[46]	training's l2: 19.0306	valid_1's l2: 48.4869
[47]	training's l2: 18.5998	valid_1's l2: 48.147
[48]	training's l2: 18.1993	valid_1's l2: 47.9369
[49]	training's l2: 17.8034	valid_1's l2: 47.8113
[50]	training's l2: 17.4247	valid_1's l2: 47.6784
[51]	training's l2: 17.0579	valid_1's l2: 47.5594
[52]	training's l2: 16.7041	valid_1's l2: 47.6436
[53]	training's l2: 16.3641	valid_1's l2: 47.4386
[54]	training's l2: 16.0406	valid_1's l2: 47.4058
[55]	training's l2: 15.7444	valid_1's l2: 47.3406
[56]	training's l2: 15.4578	valid_1's l2: 47.4111
[57]	training's l2: 15.1678	valid_1's l2: 47.3337
[58]	training's l2: 14.8767	valid_1's l2: 47.1801
[59]	training's l2: 14.6091	valid_1's l2: 47.173
[60]	training's l2: 14.3374	valid_1's l2: 47.2314
[61]	training's l2: 14.0646	valid_1's l2: 47.2778
[62

[214]	training's l2: 2.60886	valid_1's l2: 46.6768
[215]	training's l2: 2.58899	valid_1's l2: 46.6481
[216]	training's l2: 2.56351	valid_1's l2: 46.6858
[217]	training's l2: 2.54118	valid_1's l2: 46.6689
[218]	training's l2: 2.51881	valid_1's l2: 46.6675
[219]	training's l2: 2.49635	valid_1's l2: 46.6798
[220]	training's l2: 2.47414	valid_1's l2: 46.6414
[221]	training's l2: 2.45041	valid_1's l2: 46.63
[222]	training's l2: 2.43433	valid_1's l2: 46.6192
[223]	training's l2: 2.4159	valid_1's l2: 46.6332
[224]	training's l2: 2.39416	valid_1's l2: 46.5881
[225]	training's l2: 2.37427	valid_1's l2: 46.5963
[226]	training's l2: 2.35379	valid_1's l2: 46.5419
[227]	training's l2: 2.33442	valid_1's l2: 46.5368
[228]	training's l2: 2.31385	valid_1's l2: 46.5347
[229]	training's l2: 2.29507	valid_1's l2: 46.5137
[230]	training's l2: 2.27674	valid_1's l2: 46.4991
[231]	training's l2: 2.25685	valid_1's l2: 46.4629
[232]	training's l2: 2.24046	valid_1's l2: 46.4533
[233]	training's l2: 2.22245	valid

[388]	training's l2: 0.661938	valid_1's l2: 46.5693
[389]	training's l2: 0.65699	valid_1's l2: 46.5714
[390]	training's l2: 0.652211	valid_1's l2: 46.5783
[391]	training's l2: 0.648851	valid_1's l2: 46.5822
[392]	training's l2: 0.64516	valid_1's l2: 46.5891
[393]	training's l2: 0.640886	valid_1's l2: 46.5849
[394]	training's l2: 0.636286	valid_1's l2: 46.5676
[395]	training's l2: 0.631812	valid_1's l2: 46.5623
[396]	training's l2: 0.627804	valid_1's l2: 46.5667
[397]	training's l2: 0.624175	valid_1's l2: 46.5681
[398]	training's l2: 0.619661	valid_1's l2: 46.5715
[399]	training's l2: 0.616015	valid_1's l2: 46.574
[400]	training's l2: 0.611417	valid_1's l2: 46.5585
[401]	training's l2: 0.606325	valid_1's l2: 46.5588
[402]	training's l2: 0.600692	valid_1's l2: 46.5415
[403]	training's l2: 0.597116	valid_1's l2: 46.5383
[404]	training's l2: 0.593599	valid_1's l2: 46.5221
[405]	training's l2: 0.589127	valid_1's l2: 46.5177
[406]	training's l2: 0.584706	valid_1's l2: 46.5176
[407]	training'

[571]	training's l2: 0.192709	valid_1's l2: 46.4033
[572]	training's l2: 0.191481	valid_1's l2: 46.399
[573]	training's l2: 0.189975	valid_1's l2: 46.4012
[574]	training's l2: 0.188628	valid_1's l2: 46.4019
[575]	training's l2: 0.187459	valid_1's l2: 46.401
[576]	training's l2: 0.186202	valid_1's l2: 46.3986
[577]	training's l2: 0.185328	valid_1's l2: 46.3986
[578]	training's l2: 0.184056	valid_1's l2: 46.3944
[579]	training's l2: 0.182719	valid_1's l2: 46.393
[580]	training's l2: 0.181624	valid_1's l2: 46.3913
[581]	training's l2: 0.18059	valid_1's l2: 46.3883
[582]	training's l2: 0.179447	valid_1's l2: 46.3811
[583]	training's l2: 0.178571	valid_1's l2: 46.3849
[584]	training's l2: 0.177205	valid_1's l2: 46.3824
[585]	training's l2: 0.176202	valid_1's l2: 46.3808
[586]	training's l2: 0.175322	valid_1's l2: 46.3801
[587]	training's l2: 0.174255	valid_1's l2: 46.3808
[588]	training's l2: 0.17319	valid_1's l2: 46.3804
[589]	training's l2: 0.171759	valid_1's l2: 46.3801
[590]	training's 

[743]	training's l2: 0.0665249	valid_1's l2: 46.4484
[744]	training's l2: 0.0661029	valid_1's l2: 46.4485
[745]	training's l2: 0.0657311	valid_1's l2: 46.4474
[746]	training's l2: 0.0653098	valid_1's l2: 46.4447
[747]	training's l2: 0.0648833	valid_1's l2: 46.4433
[748]	training's l2: 0.0644712	valid_1's l2: 46.4455
[749]	training's l2: 0.0641349	valid_1's l2: 46.4453
[750]	training's l2: 0.0637645	valid_1's l2: 46.4471
[751]	training's l2: 0.0633672	valid_1's l2: 46.4479
[752]	training's l2: 0.0630258	valid_1's l2: 46.446
[753]	training's l2: 0.06273	valid_1's l2: 46.4455
[754]	training's l2: 0.0623872	valid_1's l2: 46.4448
[755]	training's l2: 0.0620419	valid_1's l2: 46.4439
[756]	training's l2: 0.0617407	valid_1's l2: 46.4422
[757]	training's l2: 0.0614489	valid_1's l2: 46.439
[758]	training's l2: 0.0610987	valid_1's l2: 46.4375
[759]	training's l2: 0.0607519	valid_1's l2: 46.4402
[760]	training's l2: 0.0604222	valid_1's l2: 46.4414
[761]	training's l2: 0.060185	valid_1's l2: 46.438

[914]	training's l2: 0.0238439	valid_1's l2: 46.4563
[915]	training's l2: 0.0236798	valid_1's l2: 46.4567
[916]	training's l2: 0.0235332	valid_1's l2: 46.4576
[917]	training's l2: 0.0233585	valid_1's l2: 46.4582
[918]	training's l2: 0.023213	valid_1's l2: 46.4587
[919]	training's l2: 0.0230953	valid_1's l2: 46.4583
[920]	training's l2: 0.0229533	valid_1's l2: 46.4581
[921]	training's l2: 0.0227881	valid_1's l2: 46.4592
[922]	training's l2: 0.0226167	valid_1's l2: 46.4598
[923]	training's l2: 0.0225126	valid_1's l2: 46.4578
[924]	training's l2: 0.0223602	valid_1's l2: 46.4584
[925]	training's l2: 0.0222219	valid_1's l2: 46.4601
[926]	training's l2: 0.0220973	valid_1's l2: 46.4587
[927]	training's l2: 0.0219478	valid_1's l2: 46.4596
[928]	training's l2: 0.0217931	valid_1's l2: 46.4608
[929]	training's l2: 0.0216927	valid_1's l2: 46.4609
[930]	training's l2: 0.0215691	valid_1's l2: 46.4587
[931]	training's l2: 0.0214438	valid_1's l2: 46.4593
[932]	training's l2: 0.0212987	valid_1's l2: 46

[93]	training's l2: 9.36634	valid_1's l2: 37.5992
[94]	training's l2: 9.25394	valid_1's l2: 37.6
[95]	training's l2: 9.14752	valid_1's l2: 37.6084
[96]	training's l2: 9.03747	valid_1's l2: 37.5868
[97]	training's l2: 8.93959	valid_1's l2: 37.5556
[98]	training's l2: 8.82693	valid_1's l2: 37.5969
[99]	training's l2: 8.72859	valid_1's l2: 37.5732
[100]	training's l2: 8.63127	valid_1's l2: 37.5762
[101]	training's l2: 8.53062	valid_1's l2: 37.5701
[102]	training's l2: 8.43619	valid_1's l2: 37.6021
[103]	training's l2: 8.34152	valid_1's l2: 37.5088
[104]	training's l2: 8.255	valid_1's l2: 37.4348
[105]	training's l2: 8.16023	valid_1's l2: 37.434
[106]	training's l2: 8.0802	valid_1's l2: 37.3689
[107]	training's l2: 7.98851	valid_1's l2: 37.3503
[108]	training's l2: 7.90801	valid_1's l2: 37.3187
[109]	training's l2: 7.83426	valid_1's l2: 37.2791
[110]	training's l2: 7.75728	valid_1's l2: 37.2331
[111]	training's l2: 7.67749	valid_1's l2: 37.2295
[112]	training's l2: 7.59356	valid_1's l2: 37

[272]	training's l2: 2.02119	valid_1's l2: 36.6163
[273]	training's l2: 2.00891	valid_1's l2: 36.6103
[274]	training's l2: 1.99648	valid_1's l2: 36.5887
[275]	training's l2: 1.98463	valid_1's l2: 36.5841
[276]	training's l2: 1.97075	valid_1's l2: 36.5951
[277]	training's l2: 1.95818	valid_1's l2: 36.5859
[278]	training's l2: 1.94627	valid_1's l2: 36.5788
[279]	training's l2: 1.93379	valid_1's l2: 36.5866
[280]	training's l2: 1.91782	valid_1's l2: 36.6076
[281]	training's l2: 1.90322	valid_1's l2: 36.6343
[282]	training's l2: 1.88787	valid_1's l2: 36.6268
[283]	training's l2: 1.8737	valid_1's l2: 36.6257
[284]	training's l2: 1.86092	valid_1's l2: 36.6373
[285]	training's l2: 1.84859	valid_1's l2: 36.639
[286]	training's l2: 1.83747	valid_1's l2: 36.6253
[287]	training's l2: 1.82293	valid_1's l2: 36.6288
[288]	training's l2: 1.81225	valid_1's l2: 36.6003
[289]	training's l2: 1.79929	valid_1's l2: 36.5865
[290]	training's l2: 1.78756	valid_1's l2: 36.5696
[291]	training's l2: 1.77756	vali

[438]	training's l2: 0.766348	valid_1's l2: 36.3443
[439]	training's l2: 0.762131	valid_1's l2: 36.344
[440]	training's l2: 0.758145	valid_1's l2: 36.3393
[441]	training's l2: 0.753547	valid_1's l2: 36.347
[442]	training's l2: 0.750215	valid_1's l2: 36.3455
[443]	training's l2: 0.745749	valid_1's l2: 36.3481
[444]	training's l2: 0.741207	valid_1's l2: 36.3561
[445]	training's l2: 0.737027	valid_1's l2: 36.3622
[446]	training's l2: 0.733098	valid_1's l2: 36.3589
[447]	training's l2: 0.72975	valid_1's l2: 36.3468
[448]	training's l2: 0.724803	valid_1's l2: 36.3349
[449]	training's l2: 0.720636	valid_1's l2: 36.333
[450]	training's l2: 0.7164	valid_1's l2: 36.3277
[451]	training's l2: 0.711871	valid_1's l2: 36.339
[452]	training's l2: 0.707678	valid_1's l2: 36.3242
[453]	training's l2: 0.704028	valid_1's l2: 36.3227
[454]	training's l2: 0.699026	valid_1's l2: 36.3252
[455]	training's l2: 0.69467	valid_1's l2: 36.3319
[456]	training's l2: 0.690542	valid_1's l2: 36.3328
[457]	training's l2:

[616]	training's l2: 0.315616	valid_1's l2: 36.3141
[617]	training's l2: 0.313953	valid_1's l2: 36.3199
[618]	training's l2: 0.312382	valid_1's l2: 36.3163
[619]	training's l2: 0.310802	valid_1's l2: 36.3158
[620]	training's l2: 0.309687	valid_1's l2: 36.3107
[621]	training's l2: 0.308032	valid_1's l2: 36.3116
[622]	training's l2: 0.306826	valid_1's l2: 36.3177
[623]	training's l2: 0.30527	valid_1's l2: 36.3181
[624]	training's l2: 0.303667	valid_1's l2: 36.3174
[625]	training's l2: 0.302185	valid_1's l2: 36.3177
[626]	training's l2: 0.300895	valid_1's l2: 36.3267
[627]	training's l2: 0.29954	valid_1's l2: 36.3306
[628]	training's l2: 0.298366	valid_1's l2: 36.3223
[629]	training's l2: 0.297095	valid_1's l2: 36.3187
[630]	training's l2: 0.295946	valid_1's l2: 36.3166
[631]	training's l2: 0.294729	valid_1's l2: 36.3234
[632]	training's l2: 0.29365	valid_1's l2: 36.3154
[633]	training's l2: 0.292628	valid_1's l2: 36.3138
[634]	training's l2: 0.290836	valid_1's l2: 36.3128
[635]	training'

[789]	training's l2: 0.149083	valid_1's l2: 36.3426
[790]	training's l2: 0.148501	valid_1's l2: 36.3406
[791]	training's l2: 0.147957	valid_1's l2: 36.3498
[792]	training's l2: 0.147415	valid_1's l2: 36.3469
[793]	training's l2: 0.146817	valid_1's l2: 36.3438
[794]	training's l2: 0.146286	valid_1's l2: 36.3418
[795]	training's l2: 0.145742	valid_1's l2: 36.3431
[796]	training's l2: 0.14517	valid_1's l2: 36.3438
[797]	training's l2: 0.144558	valid_1's l2: 36.3488
[798]	training's l2: 0.144047	valid_1's l2: 36.3532
[799]	training's l2: 0.143494	valid_1's l2: 36.3521
[800]	training's l2: 0.142934	valid_1's l2: 36.3487
[801]	training's l2: 0.142328	valid_1's l2: 36.3497
[802]	training's l2: 0.14164	valid_1's l2: 36.3525
[803]	training's l2: 0.141144	valid_1's l2: 36.3492
[804]	training's l2: 0.140509	valid_1's l2: 36.3468
[805]	training's l2: 0.139938	valid_1's l2: 36.3439
[806]	training's l2: 0.139286	valid_1's l2: 36.3467
[807]	training's l2: 0.138673	valid_1's l2: 36.3483
[808]	training

[963]	training's l2: 0.0751424	valid_1's l2: 36.3531
[964]	training's l2: 0.074865	valid_1's l2: 36.3534
[965]	training's l2: 0.0746076	valid_1's l2: 36.3516
[966]	training's l2: 0.0743202	valid_1's l2: 36.3531
[967]	training's l2: 0.0740493	valid_1's l2: 36.3512
[968]	training's l2: 0.0738093	valid_1's l2: 36.3493
[969]	training's l2: 0.0735726	valid_1's l2: 36.3515
[970]	training's l2: 0.0732964	valid_1's l2: 36.3467
[971]	training's l2: 0.0730658	valid_1's l2: 36.3436
[972]	training's l2: 0.0727546	valid_1's l2: 36.3416
[973]	training's l2: 0.0725116	valid_1's l2: 36.3419
[974]	training's l2: 0.0722636	valid_1's l2: 36.3463
[975]	training's l2: 0.0720122	valid_1's l2: 36.3482
[976]	training's l2: 0.0717679	valid_1's l2: 36.3464
[977]	training's l2: 0.0715056	valid_1's l2: 36.347
[978]	training's l2: 0.0712236	valid_1's l2: 36.3481
[979]	training's l2: 0.070971	valid_1's l2: 36.3482
[980]	training's l2: 0.0707544	valid_1's l2: 36.3457
[981]	training's l2: 0.0705009	valid_1's l2: 36.3

[135]	training's l2: 6.393	valid_1's l2: 33.8676
[136]	training's l2: 6.33797	valid_1's l2: 33.8911
[137]	training's l2: 6.28213	valid_1's l2: 33.9263
[138]	training's l2: 6.22058	valid_1's l2: 33.901
[139]	training's l2: 6.15827	valid_1's l2: 33.9304
[140]	training's l2: 6.09493	valid_1's l2: 33.9216
[141]	training's l2: 6.04581	valid_1's l2: 33.8839
[142]	training's l2: 5.9902	valid_1's l2: 33.8626
[143]	training's l2: 5.94717	valid_1's l2: 33.8849
[144]	training's l2: 5.89279	valid_1's l2: 33.8773
[145]	training's l2: 5.83725	valid_1's l2: 33.8941
[146]	training's l2: 5.79245	valid_1's l2: 33.8863
[147]	training's l2: 5.74308	valid_1's l2: 33.8315
[148]	training's l2: 5.68846	valid_1's l2: 33.8458
[149]	training's l2: 5.6279	valid_1's l2: 33.8775
[150]	training's l2: 5.57382	valid_1's l2: 33.8963
[151]	training's l2: 5.53056	valid_1's l2: 33.9107
[152]	training's l2: 5.48709	valid_1's l2: 33.9044
[153]	training's l2: 5.44441	valid_1's l2: 33.9404
[154]	training's l2: 5.39542	valid_1

[299]	training's l2: 1.8963	valid_1's l2: 34.0224
[300]	training's l2: 1.88459	valid_1's l2: 34.0201
[301]	training's l2: 1.8743	valid_1's l2: 34.0067
[302]	training's l2: 1.86307	valid_1's l2: 33.9847
[303]	training's l2: 1.85136	valid_1's l2: 33.9938
[304]	training's l2: 1.83904	valid_1's l2: 34.0112
[305]	training's l2: 1.82769	valid_1's l2: 34.0129
[306]	training's l2: 1.81436	valid_1's l2: 34.0048
[307]	training's l2: 1.80398	valid_1's l2: 33.9964
[308]	training's l2: 1.79167	valid_1's l2: 33.9864
[309]	training's l2: 1.78112	valid_1's l2: 33.995
[310]	training's l2: 1.76822	valid_1's l2: 33.9987
[311]	training's l2: 1.75673	valid_1's l2: 34.0104
[312]	training's l2: 1.74658	valid_1's l2: 34
[313]	training's l2: 1.73644	valid_1's l2: 34.033
[314]	training's l2: 1.72585	valid_1's l2: 34.0464
[315]	training's l2: 1.71463	valid_1's l2: 34.0471
[316]	training's l2: 1.70595	valid_1's l2: 34.0526
[317]	training's l2: 1.69415	valid_1's l2: 34.0321
[318]	training's l2: 1.68538	valid_1's l

[482]	training's l2: 0.701267	valid_1's l2: 34.2025
[483]	training's l2: 0.69819	valid_1's l2: 34.2142
[484]	training's l2: 0.694619	valid_1's l2: 34.2161
[485]	training's l2: 0.690552	valid_1's l2: 34.2146
[486]	training's l2: 0.686737	valid_1's l2: 34.2203
[487]	training's l2: 0.683172	valid_1's l2: 34.2098
[488]	training's l2: 0.67985	valid_1's l2: 34.21
[489]	training's l2: 0.677096	valid_1's l2: 34.1952
[490]	training's l2: 0.673602	valid_1's l2: 34.1964
[491]	training's l2: 0.670614	valid_1's l2: 34.2047
[492]	training's l2: 0.667497	valid_1's l2: 34.2016
[493]	training's l2: 0.663584	valid_1's l2: 34.1954
[494]	training's l2: 0.661038	valid_1's l2: 34.2003
[495]	training's l2: 0.657559	valid_1's l2: 34.1952
[496]	training's l2: 0.65473	valid_1's l2: 34.2035
[497]	training's l2: 0.651343	valid_1's l2: 34.2105
[498]	training's l2: 0.648077	valid_1's l2: 34.2162
[499]	training's l2: 0.644277	valid_1's l2: 34.2257
[500]	training's l2: 0.641589	valid_1's l2: 34.2247
[501]	training's 

[654]	training's l2: 0.316386	valid_1's l2: 34.3357
[655]	training's l2: 0.314735	valid_1's l2: 34.3339
[656]	training's l2: 0.313628	valid_1's l2: 34.3341
[657]	training's l2: 0.312518	valid_1's l2: 34.329
[658]	training's l2: 0.311005	valid_1's l2: 34.3251
[659]	training's l2: 0.309689	valid_1's l2: 34.3243
[660]	training's l2: 0.308552	valid_1's l2: 34.3259
[661]	training's l2: 0.307389	valid_1's l2: 34.33
[662]	training's l2: 0.305798	valid_1's l2: 34.3294
[663]	training's l2: 0.30425	valid_1's l2: 34.3222
[664]	training's l2: 0.302921	valid_1's l2: 34.324
[665]	training's l2: 0.301731	valid_1's l2: 34.3329
[666]	training's l2: 0.300373	valid_1's l2: 34.3422
[667]	training's l2: 0.299156	valid_1's l2: 34.3344
[668]	training's l2: 0.297698	valid_1's l2: 34.3279
[669]	training's l2: 0.296411	valid_1's l2: 34.3215
[670]	training's l2: 0.295048	valid_1's l2: 34.3243
[671]	training's l2: 0.293577	valid_1's l2: 34.3204
[672]	training's l2: 0.292377	valid_1's l2: 34.3213
[673]	training's 

[822]	training's l2: 0.155273	valid_1's l2: 34.3877
[823]	training's l2: 0.15438	valid_1's l2: 34.391
[824]	training's l2: 0.153647	valid_1's l2: 34.398
[825]	training's l2: 0.153015	valid_1's l2: 34.401
[826]	training's l2: 0.152326	valid_1's l2: 34.3989
[827]	training's l2: 0.151639	valid_1's l2: 34.399
[828]	training's l2: 0.151095	valid_1's l2: 34.3989
[829]	training's l2: 0.150506	valid_1's l2: 34.4023
[830]	training's l2: 0.149857	valid_1's l2: 34.4024
[831]	training's l2: 0.149191	valid_1's l2: 34.4006
[832]	training's l2: 0.148545	valid_1's l2: 34.3997
[833]	training's l2: 0.148052	valid_1's l2: 34.3954
[834]	training's l2: 0.147583	valid_1's l2: 34.4017
[835]	training's l2: 0.146998	valid_1's l2: 34.3996
[836]	training's l2: 0.146379	valid_1's l2: 34.3981
[837]	training's l2: 0.145676	valid_1's l2: 34.3941
[838]	training's l2: 0.145088	valid_1's l2: 34.3912
[839]	training's l2: 0.144531	valid_1's l2: 34.3922
[840]	training's l2: 0.144009	valid_1's l2: 34.3918
[841]	training's 

[990]	training's l2: 0.081415	valid_1's l2: 34.4505
[991]	training's l2: 0.0811575	valid_1's l2: 34.4478
[992]	training's l2: 0.0808755	valid_1's l2: 34.4529
[993]	training's l2: 0.0805897	valid_1's l2: 34.4538
[994]	training's l2: 0.0802835	valid_1's l2: 34.449
[995]	training's l2: 0.0800005	valid_1's l2: 34.4476
[996]	training's l2: 0.0796581	valid_1's l2: 34.4453
[997]	training's l2: 0.0794092	valid_1's l2: 34.449
[998]	training's l2: 0.0791518	valid_1's l2: 34.4489
[999]	training's l2: 0.0788603	valid_1's l2: 34.4474
[1000]	training's l2: 0.078579	valid_1's l2: 34.4491
err:  5.869337738861242
[1]	training's l2: 152.65	valid_1's l2: 154.171
[2]	training's l2: 141.468	valid_1's l2: 144.222
[3]	training's l2: 131.333	valid_1's l2: 135.123
[4]	training's l2: 122.137	valid_1's l2: 126.677
[5]	training's l2: 113.773	valid_1's l2: 118.882
[6]	training's l2: 106.165	valid_1's l2: 111.982
[7]	training's l2: 99.2118	valid_1's l2: 105.684
[8]	training's l2: 92.8158	valid_1's l2: 99.8562
[9]	t

[180]	training's l2: 4.38337	valid_1's l2: 30.1954
[181]	training's l2: 4.35142	valid_1's l2: 30.2011
[182]	training's l2: 4.31308	valid_1's l2: 30.1854
[183]	training's l2: 4.27846	valid_1's l2: 30.1723
[184]	training's l2: 4.24198	valid_1's l2: 30.193
[185]	training's l2: 4.20959	valid_1's l2: 30.1989
[186]	training's l2: 4.1791	valid_1's l2: 30.2192
[187]	training's l2: 4.15027	valid_1's l2: 30.2061
[188]	training's l2: 4.11316	valid_1's l2: 30.2405
[189]	training's l2: 4.07861	valid_1's l2: 30.2172
[190]	training's l2: 4.04844	valid_1's l2: 30.1872
[191]	training's l2: 4.0138	valid_1's l2: 30.2168
[192]	training's l2: 3.98065	valid_1's l2: 30.2315
[193]	training's l2: 3.94531	valid_1's l2: 30.2578
[194]	training's l2: 3.91378	valid_1's l2: 30.2855
[195]	training's l2: 3.88881	valid_1's l2: 30.2735
[196]	training's l2: 3.86507	valid_1's l2: 30.2552
[197]	training's l2: 3.83682	valid_1's l2: 30.282
[198]	training's l2: 3.80928	valid_1's l2: 30.2695
[199]	training's l2: 3.78086	valid_

[349]	training's l2: 1.42207	valid_1's l2: 30.3075
[350]	training's l2: 1.41281	valid_1's l2: 30.3089
[351]	training's l2: 1.4062	valid_1's l2: 30.3093
[352]	training's l2: 1.39636	valid_1's l2: 30.3061
[353]	training's l2: 1.38723	valid_1's l2: 30.2993
[354]	training's l2: 1.37911	valid_1's l2: 30.2906
[355]	training's l2: 1.37188	valid_1's l2: 30.2846
[356]	training's l2: 1.3653	valid_1's l2: 30.2837
[357]	training's l2: 1.35776	valid_1's l2: 30.2818
[358]	training's l2: 1.35059	valid_1's l2: 30.2818
[359]	training's l2: 1.34328	valid_1's l2: 30.2802
[360]	training's l2: 1.33569	valid_1's l2: 30.2737
[361]	training's l2: 1.32899	valid_1's l2: 30.2658
[362]	training's l2: 1.32199	valid_1's l2: 30.2653
[363]	training's l2: 1.31511	valid_1's l2: 30.2628
[364]	training's l2: 1.30775	valid_1's l2: 30.2573
[365]	training's l2: 1.30031	valid_1's l2: 30.2511
[366]	training's l2: 1.29189	valid_1's l2: 30.2568
[367]	training's l2: 1.28505	valid_1's l2: 30.267
[368]	training's l2: 1.27723	valid

[518]	training's l2: 0.591383	valid_1's l2: 30.1825
[519]	training's l2: 0.587864	valid_1's l2: 30.1859
[520]	training's l2: 0.584666	valid_1's l2: 30.1732
[521]	training's l2: 0.581793	valid_1's l2: 30.1731
[522]	training's l2: 0.579179	valid_1's l2: 30.1714
[523]	training's l2: 0.576048	valid_1's l2: 30.1758
[524]	training's l2: 0.574038	valid_1's l2: 30.1647
[525]	training's l2: 0.571187	valid_1's l2: 30.1629
[526]	training's l2: 0.568928	valid_1's l2: 30.1569
[527]	training's l2: 0.565566	valid_1's l2: 30.1621
[528]	training's l2: 0.562847	valid_1's l2: 30.1585
[529]	training's l2: 0.560127	valid_1's l2: 30.1556
[530]	training's l2: 0.557613	valid_1's l2: 30.1613
[531]	training's l2: 0.554451	valid_1's l2: 30.1592
[532]	training's l2: 0.551909	valid_1's l2: 30.1542
[533]	training's l2: 0.548984	valid_1's l2: 30.1617
[534]	training's l2: 0.546619	valid_1's l2: 30.1608
[535]	training's l2: 0.544473	valid_1's l2: 30.161
[536]	training's l2: 0.541627	valid_1's l2: 30.1607
[537]	trainin

[683]	training's l2: 0.279899	valid_1's l2: 30.1585
[684]	training's l2: 0.27878	valid_1's l2: 30.1559
[685]	training's l2: 0.277666	valid_1's l2: 30.1551
[686]	training's l2: 0.276685	valid_1's l2: 30.1579
[687]	training's l2: 0.275739	valid_1's l2: 30.156
[688]	training's l2: 0.274254	valid_1's l2: 30.149
[689]	training's l2: 0.272962	valid_1's l2: 30.1495
[690]	training's l2: 0.271721	valid_1's l2: 30.1534
[691]	training's l2: 0.270635	valid_1's l2: 30.1486
[692]	training's l2: 0.269423	valid_1's l2: 30.1479
[693]	training's l2: 0.268056	valid_1's l2: 30.1406
[694]	training's l2: 0.266972	valid_1's l2: 30.1477
[695]	training's l2: 0.266008	valid_1's l2: 30.1433
[696]	training's l2: 0.265097	valid_1's l2: 30.1424
[697]	training's l2: 0.26392	valid_1's l2: 30.1391
[698]	training's l2: 0.262801	valid_1's l2: 30.1436
[699]	training's l2: 0.261388	valid_1's l2: 30.1434
[700]	training's l2: 0.260405	valid_1's l2: 30.1408
[701]	training's l2: 0.259194	valid_1's l2: 30.1413
[702]	training's

[846]	training's l2: 0.145607	valid_1's l2: 30.1418
[847]	training's l2: 0.144994	valid_1's l2: 30.1368
[848]	training's l2: 0.144502	valid_1's l2: 30.1395
[849]	training's l2: 0.143915	valid_1's l2: 30.1366
[850]	training's l2: 0.143405	valid_1's l2: 30.1372
[851]	training's l2: 0.142918	valid_1's l2: 30.1373
[852]	training's l2: 0.142499	valid_1's l2: 30.1396
[853]	training's l2: 0.141836	valid_1's l2: 30.1376
[854]	training's l2: 0.141381	valid_1's l2: 30.1351
[855]	training's l2: 0.140847	valid_1's l2: 30.1359
[856]	training's l2: 0.140359	valid_1's l2: 30.1363
[857]	training's l2: 0.139835	valid_1's l2: 30.1366
[858]	training's l2: 0.139249	valid_1's l2: 30.1366
[859]	training's l2: 0.138685	valid_1's l2: 30.1347
[860]	training's l2: 0.138191	valid_1's l2: 30.1368
[861]	training's l2: 0.137604	valid_1's l2: 30.1377
[862]	training's l2: 0.137081	valid_1's l2: 30.1405
[863]	training's l2: 0.136537	valid_1's l2: 30.1399
[864]	training's l2: 0.135957	valid_1's l2: 30.1426
[865]	traini

[17]	training's l2: 54.5837	valid_1's l2: 58.424
[18]	training's l2: 51.8973	valid_1's l2: 56.4096
[19]	training's l2: 49.442	valid_1's l2: 54.5683
[20]	training's l2: 47.1482	valid_1's l2: 52.7414
[21]	training's l2: 45.0576	valid_1's l2: 51.2486
[22]	training's l2: 43.0903	valid_1's l2: 49.9121
[23]	training's l2: 41.2544	valid_1's l2: 48.5686
[24]	training's l2: 39.5782	valid_1's l2: 47.4812
[25]	training's l2: 38.0268	valid_1's l2: 46.5374
[26]	training's l2: 36.6031	valid_1's l2: 45.5806
[27]	training's l2: 35.2542	valid_1's l2: 44.806
[28]	training's l2: 33.9699	valid_1's l2: 43.971
[29]	training's l2: 32.7915	valid_1's l2: 43.2278
[30]	training's l2: 31.709	valid_1's l2: 42.5498
[31]	training's l2: 30.5966	valid_1's l2: 41.8098
[32]	training's l2: 29.6738	valid_1's l2: 41.4257
[33]	training's l2: 28.6851	valid_1's l2: 40.7804
[34]	training's l2: 27.7801	valid_1's l2: 40.1437
[35]	training's l2: 26.9881	valid_1's l2: 39.5836
[36]	training's l2: 26.258	valid_1's l2: 39.1372
[37]	t

[201]	training's l2: 3.68556	valid_1's l2: 33.2329
[202]	training's l2: 3.65558	valid_1's l2: 33.2587
[203]	training's l2: 3.62549	valid_1's l2: 33.2123
[204]	training's l2: 3.59296	valid_1's l2: 33.2064
[205]	training's l2: 3.55882	valid_1's l2: 33.196
[206]	training's l2: 3.53704	valid_1's l2: 33.2097
[207]	training's l2: 3.50775	valid_1's l2: 33.2246
[208]	training's l2: 3.47877	valid_1's l2: 33.2252
[209]	training's l2: 3.45447	valid_1's l2: 33.2106
[210]	training's l2: 3.43619	valid_1's l2: 33.1932
[211]	training's l2: 3.40958	valid_1's l2: 33.1933
[212]	training's l2: 3.38482	valid_1's l2: 33.1885
[213]	training's l2: 3.36545	valid_1's l2: 33.1845
[214]	training's l2: 3.33976	valid_1's l2: 33.176
[215]	training's l2: 3.3075	valid_1's l2: 33.1967
[216]	training's l2: 3.2889	valid_1's l2: 33.2068
[217]	training's l2: 3.26434	valid_1's l2: 33.1736
[218]	training's l2: 3.24056	valid_1's l2: 33.1516
[219]	training's l2: 3.21612	valid_1's l2: 33.13
[220]	training's l2: 3.19399	valid_1'

[385]	training's l2: 1.11045	valid_1's l2: 32.977
[386]	training's l2: 1.10507	valid_1's l2: 32.9751
[387]	training's l2: 1.09921	valid_1's l2: 32.9721
[388]	training's l2: 1.09238	valid_1's l2: 32.9677
[389]	training's l2: 1.08561	valid_1's l2: 32.9596
[390]	training's l2: 1.08041	valid_1's l2: 32.9633
[391]	training's l2: 1.07463	valid_1's l2: 32.968
[392]	training's l2: 1.06896	valid_1's l2: 32.9707
[393]	training's l2: 1.06337	valid_1's l2: 32.9706
[394]	training's l2: 1.05702	valid_1's l2: 32.9763
[395]	training's l2: 1.051	valid_1's l2: 32.9696
[396]	training's l2: 1.04603	valid_1's l2: 32.9596
[397]	training's l2: 1.04028	valid_1's l2: 32.9631
[398]	training's l2: 1.03486	valid_1's l2: 32.9593
[399]	training's l2: 1.02895	valid_1's l2: 32.9613
[400]	training's l2: 1.0238	valid_1's l2: 32.959
[401]	training's l2: 1.01806	valid_1's l2: 32.9648
[402]	training's l2: 1.01229	valid_1's l2: 32.9596
[403]	training's l2: 1.00687	valid_1's l2: 32.9606
[404]	training's l2: 1.00169	valid_1'

[545]	training's l2: 0.483973	valid_1's l2: 32.8854
[546]	training's l2: 0.48146	valid_1's l2: 32.8863
[547]	training's l2: 0.478903	valid_1's l2: 32.8896
[548]	training's l2: 0.476583	valid_1's l2: 32.8975
[549]	training's l2: 0.474707	valid_1's l2: 32.9005
[550]	training's l2: 0.472646	valid_1's l2: 32.9067
[551]	training's l2: 0.469884	valid_1's l2: 32.9091
[552]	training's l2: 0.467447	valid_1's l2: 32.9068
[553]	training's l2: 0.465249	valid_1's l2: 32.8976
[554]	training's l2: 0.463292	valid_1's l2: 32.892
[555]	training's l2: 0.461165	valid_1's l2: 32.8864
[556]	training's l2: 0.458738	valid_1's l2: 32.8865
[557]	training's l2: 0.456362	valid_1's l2: 32.8882
[558]	training's l2: 0.454254	valid_1's l2: 32.8901
[559]	training's l2: 0.452083	valid_1's l2: 32.8863
[560]	training's l2: 0.449996	valid_1's l2: 32.8871
[561]	training's l2: 0.448098	valid_1's l2: 32.8885
[562]	training's l2: 0.446105	valid_1's l2: 32.8893
[563]	training's l2: 0.44439	valid_1's l2: 32.8887
[564]	training'

[730]	training's l2: 0.21168	valid_1's l2: 32.9191
[731]	training's l2: 0.210585	valid_1's l2: 32.9198
[732]	training's l2: 0.209691	valid_1's l2: 32.9182
[733]	training's l2: 0.208818	valid_1's l2: 32.9178
[734]	training's l2: 0.207872	valid_1's l2: 32.9164
[735]	training's l2: 0.207193	valid_1's l2: 32.9177
[736]	training's l2: 0.20622	valid_1's l2: 32.9188
[737]	training's l2: 0.205425	valid_1's l2: 32.9208
[738]	training's l2: 0.204489	valid_1's l2: 32.9198
[739]	training's l2: 0.203567	valid_1's l2: 32.9187
[740]	training's l2: 0.202805	valid_1's l2: 32.9195
[741]	training's l2: 0.201858	valid_1's l2: 32.9177
[742]	training's l2: 0.200851	valid_1's l2: 32.9174
[743]	training's l2: 0.199976	valid_1's l2: 32.9177
[744]	training's l2: 0.199109	valid_1's l2: 32.9185
[745]	training's l2: 0.198268	valid_1's l2: 32.9198
[746]	training's l2: 0.197317	valid_1's l2: 32.9138
[747]	training's l2: 0.196617	valid_1's l2: 32.9134
[748]	training's l2: 0.195734	valid_1's l2: 32.9135
[749]	training

[913]	training's l2: 0.103281	valid_1's l2: 32.8952
[914]	training's l2: 0.102892	valid_1's l2: 32.8984
[915]	training's l2: 0.102514	valid_1's l2: 32.8984
[916]	training's l2: 0.102038	valid_1's l2: 32.8939
[917]	training's l2: 0.101672	valid_1's l2: 32.8937
[918]	training's l2: 0.101248	valid_1's l2: 32.8935
[919]	training's l2: 0.100799	valid_1's l2: 32.8924
[920]	training's l2: 0.100426	valid_1's l2: 32.8908
[921]	training's l2: 0.100084	valid_1's l2: 32.8916
[922]	training's l2: 0.0997057	valid_1's l2: 32.8927
[923]	training's l2: 0.0993588	valid_1's l2: 32.8955
[924]	training's l2: 0.0990076	valid_1's l2: 32.8958
[925]	training's l2: 0.0986597	valid_1's l2: 32.8946
[926]	training's l2: 0.0982498	valid_1's l2: 32.8957
[927]	training's l2: 0.0978666	valid_1's l2: 32.8947
[928]	training's l2: 0.0974894	valid_1's l2: 32.8964
[929]	training's l2: 0.0970969	valid_1's l2: 32.8994
[930]	training's l2: 0.0967197	valid_1's l2: 32.8972
[931]	training's l2: 0.0964121	valid_1's l2: 32.8979
[9

In [121]:
score

[6.020986223436029,
 6.015411374882773,
 6.816948350833014,
 6.028448071271268,
 5.869337738861242,
 5.4905455006475865,
 5.737646484967558]

In [122]:
np.mean(score)

5.997046249271352

> Using Cross-validation and hyperparameter tuning, we were able to improve our baseline's model performance from an RMSE of 6.1 to an RMSE of 5.99.